### ISSUE TITLE
The order of weights is changed for array of layers inside sub classed model

### ISSUE THREAD

**Issue by RanjanBalappa** [tensorflow/tensorflow#41150]
The order of  moving variance and moving mean weights(BatchNormalization layer) changes when array of layers is created inside sub classed model

````python
import tensorflow as tf
from tensorflow.keras.layers import *
print(tf.__version__)

class A(tf.keras.layers.Layer):
    def  __init__(self,  name):
      super(A, self).__init__(name=name)
      self.conv = Conv2D(32, 3)
      self.bn = BatchNormalization()  
    def call(self,  x):
      x = self.conv(x)
      x = self.bn(x)
      return x

class B(tf.keras.Model):
    def __init__(self):
        super(B, self).__init__()
        self.blocks = []
        for i in range(3):
          self.blocks.append(A(f'a{i}'))    
    
    def call(self, x):
        for block in self.blocks:
          x = block(x)
        return x

b = B()
b.build((1, 128, 128, 3))
print([weight.name for weight in b.weights])

#The output is as below
'a0/conv2d/kernel:0',
 'a0/conv2d/bias:0',
 'a0/batch_normalization/gamma:0',
 'a0/batch_normalization/beta:0',
 'a1/conv2d_1/kernel:0',
 'a1/conv2d_1/bias:0',
 'a1/batch_normalization_1/gamma:0',
 'a1/batch_normalization_1/beta:0',
 'a2/conv2d_2/kernel:0',
 'a2/conv2d_2/bias:0',
 'a2/batch_normalization_2/gamma:0',
 'a2/batch_normalization_2/beta:0',
 'a0/batch_normalization/moving_mean:0',
 'a0/batch_normalization/moving_variance:0',
 'a1/batch_normalization_1/moving_mean:0',
 'a1/batch_normalization_1/moving_variance:0',
 'a2/batch_normalization_2/moving_mean:0',
 'a2/batch_normalization_2/moving_variance:0'
